### PR TITLE
[SPARK-22624][PYSPARK] Expose range partitioning shuffle introduced by spark-22614

### PR DESCRIPTION
**Pull request by xubo245**
## What changes were proposed in this pull request?

 Expose range partitioning shuffle introduced by spark-22614

## How was this patch tested?

Unit test in dataframe.py

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by xubo245** [comment]
@gatorsmile  please review it.

**Comment by gatorsmile** [comment]
cc @ueshin @HyukjinKwon

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -667,6 +667,55 @@ def repartition(self, numPartitions, *cols):
...
 
+    @since("2.3.0")
+    def repartitionByRange(self, numPartitions, *cols, **kwargs):
+        """
+        Returns a new :class:`DataFrame` partitioned by the given partitioning expressions. The
+        resulting DataFrame is range partitioned.
+
+        ``numPartitions`` can be an int to specify the target number of partitions or a Column.
+        If it is a Column, it will be used as the first partitioning column. If not specified,
+        the default number of partitions is used.
```
Comment (gatorsmile) [inline python/pyspark/sql/dataframe.py]: Also add a test case for this?

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -667,6 +667,55 @@ def repartition(self, numPartitions, *cols):
...
+        >>> data.rdd.getNumPartitions()
+        3
+        """
+        if isinstance(numPartitions, int):
+            if len(cols) == 0:
+                return ValueError("At least one partition-by expression must be specified.")
+            else:
+                return DataFrame(
+                    self._jdf.repartitionByRange(numPartitions, self._jcols(*cols)), self.sql_ctx)
+        else:
```
Comment (gatorsmile) [inline python/pyspark/sql/dataframe.py]: It sounds like we missing the case when `numPartitions ` is not provided. Please check the above implementation of `repartition`?

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -667,6 +667,55 @@ def repartition(self, numPartitions, *cols):
         else:
             raise TypeError("numPartitions should be an int or Column")
 
+    @since("2.3.0")
+    def repartitionByRange(self, numPartitions, *cols, **kwargs):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: Mind if I ask why we added `**kwargs`?

File: python/pyspark/sql/dataframe.py [hunk 4]
```diff
@@ -667,6 +667,55 @@ def repartition(self, numPartitions, *cols):
...
+        ``numPartitions`` can be an int to specify the target number of partitions or a Column.
+        If it is a Column, it will be used as the first partitioning column. If not specified,
+        the default number of partitions is used.
+
+        At least one partition-by expression must be specified.
+        When no explicit sort order is specified, "ascending nulls first" is assumed.
+
+        >>> df.repartitionByRange(2, "age").rdd.getNumPartitions()
+        2
+        >>> data = df.union(df).repartition(1, "age")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: I think we don't have to test `repartition` within `repartitionByRange`'s doctest but leave

``  .. seealso:: :meth:`repartition` ``

File: python/pyspark/sql/dataframe.py (hunk 2 above)
Comment (xubo245) [reply python/pyspark/sql/dataframe.py]: throw  typeError， how to handle in test case?

Comment (xubo245) [reply python/pyspark/sql/dataframe.py]: ok,done, please review

File: python/pyspark/sql/dataframe.py (hunk 1 above)
Comment (xubo245) [reply python/pyspark/sql/dataframe.py]: ok, done, please review

**Comment by xubo245** [comment]
@gatorsmile @HyukjinKwon Please review it again, thanks

File: python/pyspark/sql/dataframe.py [hunk 5]
```diff
@@ -667,6 +667,92 @@ def repartition(self, numPartitions, *cols):
...
+        |  2|Alice|
+        |  5|  Bob|
+        |  5|  Bob|
+        +---+-----+
+        >>> data.rdd.getNumPartitions()
+        3
+        >>> data = data.repartitionByRange("age")
+        >>> data.rdd.getNumPartitions()
+        3
+        >>> data2 = df.union(df).union(df)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: I think we don't need `union` too.

File: python/pyspark/sql/dataframe.py [hunk 6]
```diff
@@ -667,6 +667,92 @@ def repartition(self, numPartitions, *cols):
...
+        >>> df.show()
+        +---+-----+
+        |age| name|
+        +---+-----+
+        |  2|Alice|
+        |  5|  Bob|
+        +---+-----+
+        >>> df.repartitionByRange(1, "age").rdd.getNumPartitions()
+        1
+        >>> data = df.union(df)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: ditto - I think `repartitionByRange` doctest should focus on `repartitionByRange`.

File: python/pyspark/sql/dataframe.py (hunk 5 above)
Comment (xubo245) [reply python/pyspark/sql/dataframe.py]: How to test data after repartitionByRange("age")

File: python/pyspark/sql/dataframe.py (hunk 6 above)
Comment (xubo245) [reply python/pyspark/sql/dataframe.py]: How to test data after repartitionByRange("age")?

File: python/pyspark/sql/dataframe.py (hunk 5 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/dataframe.py]: If it's a test purpose only, let's put this in `pyspark/sql/tests.py`.

Comment (xubo245) [reply python/pyspark/sql/dataframe.py]: ok, remove union

File: python/pyspark/sql/dataframe.py [hunk 7]
```diff
@@ -667,6 +667,51 @@ def repartition(self, numPartitions, *cols):
...
+            if len(cols) == 0:
+                return ValueError("At least one partition-by expression must be specified.")
+            else:
+                return DataFrame(
+                    self._jdf.repartitionByRange(numPartitions, self._jcols(*cols)), self.sql_ctx)
+        elif isinstance(numPartitions, (basestring, Column)):
+            cols = (numPartitions,) + cols
+            return DataFrame(self._jdf.repartitionByRange(self._jcols(*cols)), self.sql_ctx)
+        else:
+            raise TypeError("numPartitions should be an int or Column")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: `an int or Column` -> `an int, string or Column`

File: python/pyspark/sql/dataframe.py [hunk 8]
```diff
@@ -667,6 +667,51 @@ def repartition(self, numPartitions, *cols):
         else:
             raise TypeError("numPartitions should be an int or Column")
 
+    @since("2.3.0")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: Hm, I prefer 2.4.0 but am okay with 2.3.0.

**Review by HyukjinKwon — APPROVED** [review]
LGTM

File: python/pyspark/sql/dataframe.py (hunk 7 above)
Comment (xubo245) [reply python/pyspark/sql/dataframe.py]: ok, done

File: python/pyspark/sql/dataframe.py (hunk 8 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/dataframe.py]: Sorry, shall we just target 2.4.0?

Comment (xubo245) [reply python/pyspark/sql/dataframe.py]: ok, rename to 2.4.0, done

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**